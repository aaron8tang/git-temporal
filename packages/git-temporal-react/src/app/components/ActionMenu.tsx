import React from 'react';
import { style } from 'app/styles';
import { ToggleButton } from 'app/components/ToggleButton';
import { Popup } from 'app/components/Popup';

export interface ActionMenuProps {
  // The children are the menu content
  children: string | JSX.Element | JSX.Element[];
  style?: string | object;
}
const initialState = {
  menuOpen: false,
};
type ActionMenuState = Readonly<typeof initialState>;

const containerStyle = {
  overflow: 'visible',
  height: 16,
  width: 28,
};

const buttonStyle = {
  _extends: 'h2Text',
  height: 29,
  width: 14,
  margin: 0,
  padding: '0px 5px',
};

const menuStyle = {
  right: 0,
  marginRight: 0,
  minWidth: 180,
};

export class ActionMenu extends React.Component<
  ActionMenuProps,
  ActionMenuState
> {
  readonly state: ActionMenuState = initialState;

  render() {
    return (
      <div style={style(this.props.style, containerStyle)}>
        <ToggleButton
          isSelected={this.state.menuOpen}
          onClick={this.onToggleClick}
          style={buttonStyle}
        >
          ...
        </ToggleButton>
        <Popup
          style={menuStyle}
          isOpen={this.state.menuOpen}
          onClose={this.onPopupClose}
        >
          {this.props.children}
        </Popup>
      </div>
    );
  }

  onToggleClick = () => {
    this.setState({ menuOpen: !this.state.menuOpen });
  };
  onPopupClose = () => {
    this.setState({ menuOpen: false });
  };
}
