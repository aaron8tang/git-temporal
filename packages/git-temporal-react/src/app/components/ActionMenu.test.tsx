import React from 'react';
import { shallow } from 'enzyme';

import { ActionMenu } from './ActionMenu';

describe('components/ActionMenu', () => {
  describe('when rendered with text child', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallow(
        <ActionMenu isMenuOpen={true} onMenuToggle={jest.fn()}>
          this is a menu (or anything really)
        </ActionMenu>
      );
    });

    test('it should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
