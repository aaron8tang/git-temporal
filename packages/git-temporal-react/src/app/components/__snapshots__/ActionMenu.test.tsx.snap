// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ActionMenu when rendered with text child it should match snapshot 1`] = `
<div
  style={
    Object {
      "height": 16,
      "overflow": "visible",
      "width": 28,
    }
  }
>
  <ToggleButton
    isSelected={true}
    onClick={[MockFunction]}
    style={
      Object {
        "_extends": Object {
          "_extends": "h2Text",
          "height": 29,
          "margin": 0,
          "padding": "0px 5px",
          "width": 14,
        },
        "borderBottomLeftRadius": "0px",
        "borderBottomRightRadius": "0px",
      }
    }
  >
    ...
  </ToggleButton>
  <Popup
    isOpen={true}
    onClose={[MockFunction]}
    style={
      Object {
        "border": "solid 2px @colors.selectable",
        "borderTopRightRadius": "0px",
        "marginRight": 0,
        "minWidth": 180,
        "right": "2px",
      }
    }
  >
    this is a menu (or anything really)
  </Popup>
</div>
`;
